projects/js-toolkit/packages/js-api/data-set/index.ts

```diff
@@ -19,26 +21,31 @@ export interface FDSCellRenderer {
 	(args: FDSCellRendererArgs): HTMLElement;
 }
 
-export interface FDSFilterData {
-	id: string;
-	odataFilterString: string;
-	selectedData: string;
+// Frontend data set filter
+
+export interface FDSFilterData<T> {
+	selectedData: T;
+}
+
+export interface FDSFilterHTMLElementBuilderArgs<T> {
+	filter: FDSFilterData<T>;
+	setFilter: (partialFilter: Partial<FDSFilterData<T>>) => void;
 }
 
-export interface FDSFilterArgs {
-	filter: FDSFilterData;
-	setFilter: (val: Partial<FDSFilterData>) => void;
+export interface FDSFilterHTMLElementBuilder<T> {
+	(args: FDSFilterHTMLElementBuilderArgs<T>): HTMLElement;
 }
 
-export interface FDSFilter {
-	(args: FDSFilterArgs): HTMLElement;
+export interface FDSFilterODataQueryBuilder<T> {
+	(selectedData: T): string;
 }
 
-export interface FDSFilterHTMLElementBuilderArgs {
-	filter: FDSFilterData;
-	setFilter: (val: Partial<FDSFilterData>) => void;
+export interface FDSFilterDescriptionBuilder<T> {
+	(selectedData: T): string;
 }
 
-export interface FDSFilterHTMLElementBuilder {
-	(args: FDSFilterHTMLElementBuilderArgs): HTMLElement;
+export interface FDSFilter<T> {
+	buildFilterDescription: FDSFilterDescriptionBuilder<T>;
+	buildHTMLElement: FDSFilterHTMLElementBuilder<T>;
+	buildODataQuery: FDSFilterODataQueryBuilder<T>;
```

Comment:
This is basically a Renderer. An object containing all necessary information to render something. We know that [the CX renderer](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/renderer/types.ts#L10-L16) should have an `htmlElementBuilder`. In case of filters, it requires two more properties. This, to me, implies something like this:
```ts
export interface IClientExtensionRenderer {
	externalReferenceCode?: string;
	htmlElementBuilder?: IHTMLElementBuilder;
	name?: string;
	type: 'clientExtension';
	url?: string;
}

export interface FDSFilterClientExtensionRenderer extends IClientExtensionRenderer {
	descriptionBuilder: FDSFilterDescriptionBuilder<T>;
    htmlElementBuilder: FDSFilterHTMLElementBuilder<T>;
    oDataQueryBuilder: FDSFilterODataQueryBuilder<T>;
}
```

Reply:
This example shows what I said below: why should the `IClientExtensionRenderer` type be in a spec like `@liferay/js-api` at all? It doesn't make sense...

```typescript
export interface IClientExtensionRenderer {
    externalReferenceCode?: string;
    htmlElementBuilder?: IHTMLElementBuilder;
    name?: string;
    type: 'clientExtension';
    url?: string;
}
```

The customer doesn't need to care about `externalReferenceCode` or `type`, for example. We want to hide that from him.

Reply:
Yeah, you definitely have a good point there!

With that in mind, my only nit on this code would be names of properties, for consistency. Could we have `descriptionBuilder` instead of `buildFilterDescription`?

Reply:
Done -> https://github.com/liferay/liferay-frontend-projects/pull/1172/commits/f9e7f8192683329d997ed9d39a49c22dcb01b7bf